python/tvm/script/parser/relax/parser.py

```diff
@@ -274,7 +274,21 @@ def post_visit_local_function(self: Parser, node: doc.Expr) -> None:
 @dispatch.register(token="relax", type_name="Expr")
 def visit_expr_stmt(self: Parser, node: doc.Expr) -> None:
     value = self.eval_expr(node.value)
-    if value is not None:
+    if isinstance(value, relax.Expr):
+        var = R.emit(value)
+        IRBuilder.name("_", var)
+        is_void_value = (
+            isinstance(var.struct_info, relax.TupleStructInfo) and len(var.struct_info.fields) == 0
+        )
+
+        if not is_void_value:
+            self.report_error(
+                node,
+                f"Non-void relax expressions must be bound to a variable, "
+                f"but expression of type {var.struct_info} was used as a statement.",
+            )
```

Comment:
I wonder if we should even have this as a rule. Why not let users evaluate expressions without binding them regardless of their return type?

Reply:
At the moment, because I wanted to make the minimal change that would support common cases.  I think it would be good to remove the restriction altogether, but for the first step, I wanted to make the restriction be explicit.

There's a couple of concerns I could see with allowing non-void return value to be implicitly ignored.

* Prevent accidentally unused values.  If two IRModule instances An in-place operator that performs `a = a + b` may be represented as `cls.add(a, b)`.  This woul
* Round-trip TVMScript -> Relax -> TVMScript without a pre-processing pass.  Checking if a value has void type can done while printing the IR.  Checking whether a non-void variable could be omitted would require a pre-processing step to find any downstream users.

I don't think either of those are definitive arguments, but I figured I'd handle the unambiguous beneficial cases first, with a follow-up PR to relax the restriction.

Reply:
Those are good points. For the first case, we could have a warning (as C and other languages do with the right settings) for ignoring a return value. The second one is an interesting issue. I think it suggests that expecting an exact textual match for the parser roundtripping is too strict of a criterion for this situation, since a "statement" can always be written as `_ = ...` and it would be a choice as to whether to write it that way or use the friendlier syntax. It would make it harder to write automatic tests, true. For a systematic solution, maybe we could formalize the idea of a desugaring step for testing purposes?